### PR TITLE
Cross-Validation & Regularization Technique Implemented

### PR DESCRIPTION
**Pull request by Sanjeev-Kumar78**
## Description
File: `main.py`

- Added K-Fold Cross-Validation:

  - Implemented K-fold cross-validation to train the model on different data splits.
  - Split the dataset into K folds and trained the model on K-1 folds while validating on the remaining fold.
  - Added L2 Weight Decay to Optimizer: Updated the optimizer to include L2 weight decay. 
  - ```optimizer = torch.optim.AdamW(unet.parameters(), lr=1e-4, weight_decay=1e-5)```

File: `lora.py`
- Added Dropout to LoRA Layers:

  - Modified the LoRALayer class to include a dropout layer.
  - Updated the forward method in LoRALayer to apply dropout to the input tensor x.
  - Modified the LoRALinear class to pass the dropout rate to the LoRALayer.

File: `train.py`

- Implemented Early Stopping:

  - Added early stopping based on validation loss to prevent overfitting and optimize training time.
  - Tracked the validation loss and stopped training if the loss did not improve after a set number of epochs.
---
Fixes # (#9 #8 #7 ) 
### Type of Change

Please delete options that are not relevant.

- [x] New feature (non-breaking change which adds functionality)
- [x] This change requires a documentation update
- [ ] Other (please specify):

### Checklist:

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] New and existing unit tests pass locally with my changes

#### Additional Information 
- Name: Sanjeev Kumar
- Email:2405605@kiit.ac.in
- Roll No:2405605
- Branch: CSE

@sohambuilds

### CODE REVIEW

**Comment by rycerzes** [comment]
@sohambuilds please review this

**Review by sohambuilds — CHANGES_REQUESTED** [review]
Please add 10-15 generation samples post fine-tuning within your directory, It is suggested but not compulsory, to use another set of images for your data(could be any character or object), and finetune on that. 

Ensure that based on the number of images in the dataset, you adjust your learning rate. 1e-5 is recommended for small datasets. of 10-20 images. 1e-4 is not appropriate.

**Comment by sohambuilds** [comment]
Running on colab should be possible, as we have tried it. If there is a specific issue that you need to troubleshoot, you may join the WhatsApp group for the ML contributors: https://chat.whatsapp.com/Kx8okfEdirALcC8UeFtl5j 

Also, do note that implementing early stopping for few shot learning(very few samples) is not desirable.

**Comment by sohambuilds** [comment]
Please run the following in your colab runtime:
`!pip uninstall -y tensorflow
!pip install tensorflow-cpu
Runtime > Restart runtime 
`

**Comment by sohambuilds** [comment]
Please commit the updated code, and your dataset, if you're using a new one. You can add the samples later, I will merge it for now.

**Comment by sohambuilds** [comment]
Merging now. Sample generation is not proper. need to check.
